lib/tree.rb

```diff
@@ -16,51 +16,93 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(logn) 
+  # Space Complexity: O(n)
+  def add(key, value, current=@root)
```

Comment:
The space/time complexity is O(log n) if the tree is balanced and O(n) if it's not.